library/src/scala/tasty/util/ShowSourceCode.scala

```diff
@@ -0,0 +1,751 @@
+        printTree(arg)
+
+      case SpecialOp("throw", expr :: Nil) =>
+        this += "throw "
+        printTree(expr)
+
+      case Term.Apply(fn, args) =>
+        printTree(fn)
+        this += "("
+        printTrees(args, ", ")
+        this += ")"
+
+      case Term.TypeApply(fn, args) =>
+        printTree(fn)
+        this += "["
+        printTypeTrees(args, ", ")
+        this += "]"
+
+      case Term.Super(qual, tptOpt) =>
+        printTree(qual)
+        this += ".super"
+        // TODO use tptOpt?
+
+      case Term.Typed(term, tpt) =>
+        this += "("
+        printTree(term)
+        this += ": "
+        printTypeTree(tpt)
+        this += ")"
+
+      case Term.Assign(lhs, rhs) =>
+        printTree(lhs)
+        this += " = "
+        printTree(rhs)
+
+      case Term.Block(stats, expr) =>
+        expr match {
+          case Term.Lambda(_, _) =>
+            // Decompile lambda from { def annon$(...) = ...; closure(annon$, ...)}
+            val DefDef(_, _, args :: Nil, _, Some(rhs)) :: Nil = stats
+            this += "("
+            printArgsDefs(args)
+            this += " => "
+            printTree(rhs)
+            this += ")"
+
+          case Term.Apply(Term.Ident("while$"), _) =>
+            val DefDef("while$", _, _, _, Some(Term.If(cond, Term.Block(body :: Nil, _), _))) = stats.head
+            this += "while ("
+            printTree(cond)
+            this += ") "
+            printTree(body)
+
+          case Term.Apply(Term.Ident("doWhile$"), _) =>
+            val DefDef("doWhile$", _, _, _, Some(Term.Block(List(body), Term.If(cond, _, _)))) = stats.head
+            this += "do "
+            printTree(body)
+            this += " while ("
+            printTree(cond)
+            this += ")"
+
+          case _ =>
+            this += "{"
+            indented {
+              if (!stats.isEmpty) {
+                this += lineBreak()
+                printTrees(stats, lineBreak())
+              }
+              this += lineBreak()
+              printTree(expr)
+            }
+            this += lineBreak() += "}"
+        }
+
+      case Term.Inlined(call, bindings, expansion) =>
+        sb.append("{ // inlined")
+        indented {
+          if (!bindings.isEmpty) {
+            this += lineBreak()
+            printTrees(bindings, lineBreak())
+          }
+          this += lineBreak()
+          printTree(expansion)
+        }
+        this += lineBreak() += "}"
+
+      case Term.Lambda(meth, tpt) =>
+        // Printed in Term.Block branch
+        this
+
+      case Term.If(cond, thenp, elsep) =>
+        this += "if ("
+        printTree(cond)
+        this += ") "
+        printTree(thenp)
+        this+= " else "
+        printTree(elsep)
+
+      case Term.Match(selector, cases) =>
+        printTree(selector)
+        this += " match {"
+        indented {
+          this += lineBreak()
+          printCases(cases, lineBreak())
+        }
+        this += lineBreak() += "}"
```

Comment:
Do we have a test for partial functions?

Another tricky issue is extractors with implicits. Maybe add a TODO to support later.

Reply:
Not yet, I will add it later. The idea is to start testing all the files in `test/pos` and `test/run`.